misk-zookeeper/src/test/kotlin/misk/clustering/zookeeper/ZkTestModule.kt

```diff
@@ -29,20 +30,31 @@ internal class ZkTestModule : KAbstractModule() {
     install(ZookeeperModule())
   }
 
-  @Singleton
-  class StartZookeeperService : AbstractIdleService(), DependentService {
+  @Singleton class StartZookeeperService : AbstractIdleService(), DependentService {
     override val consumedKeys: Set<Key<*>> = setOf()
     override val producedKeys: Set<Key<*>> = setOf(startZkServiceKey)
-    private val sharedZookeeper = EmbeddedZookeeper(zkPortKey)
 
     override fun startUp() {
-      log.info { "starting zk instance " }
-      sharedZookeeper.start()
+      if (hasStarted.compareAndSet(false, true)) {
+        log.info { "starting zk instance" }
+        sharedZookeeper.start()
+        Runtime.getRuntime().addShutdownHook(Thread {
+          log.info { "stopping zk instance" }
+          sharedZookeeper.stop()
+        })
+      } else {
+        log.info { "zk instance already running" }
+      }
     }
 
     override fun shutDown() {
-      log.info { "stopping zk instance" }
-      sharedZookeeper.stop()
+      // Shutdown happens as a runtime shutdown hook
```

Comment:
This is so that ZK doesn't shutdown prematurely before all tests are done running?

Reply:
before this change, every test spun up the service. With this change, all tests within the same vm will reuse the same zk service